test/metabase/lib/test_util/generators.cljc

```diff
@@ -194,6 +195,63 @@
             (is (= (first filter-clause)
                    (:short (lib/filter-operator after stage-number (last after-filters)))))))))))
 
+;; Expressions ===================================================================================
+;; We only support a few basic expressions for now. It would be good to exercise all the expression types eventually,
+;; but the main objective here is to generate *some* expressions so they can be consumed by filters, aggregations, etc.
+;; since that's a major bug source.
+(add-step {:kind :expression})
+
+(defn- gen-expression:number [column]
+  (lib/+ column 1))
+
+(defn- gen-expression:string [column]
+  (lib/concat column "__concat"))
+
+(defn- gen-expression [columns]
+  (let [numbers (map #(vector gen-expression:number %) (filter lib.types.isa/number? columns))
+        strings (map #(vector gen-expression:string %) (filter lib.types.isa/string? columns))
+        [f col] (gen.u/choose (concat numbers strings))]
+    (f col)))
+
+(def ^:private identifier-chars-initial
+  (str "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+       "abcdefghijklmnopqrstuvwxyz"
+       "_"))
+
+(def ^:private identifier-chars
+  (str identifier-chars-initial "0123456789"))
+
+(def ^:private expr-name-chars
+  (str identifier-chars " []<>+!@$%&*^()"))
+
+(defn- gen-expression-name []
+  (let [len (gen.u/choose (range 3 30))]
+    (apply str (repeatedly len #(gen.u/choose expr-name-chars)))))
+
+(defmethod next-steps* :expression [query _expression]
+  (let [stage-number (choose-stage query)
+        ;; Always adding at the end for now. Editing will come later.
+        expr-pos     (count (lib/expressions query stage-number))]
+    (when-let [expr-clause (gen-expression (lib/expressionable-columns query stage-number expr-pos))]
+      [:expression stage-number (gen-expression-name) expr-clause])))
+
+(defmethod run-step* :expression [query [_expression stage-number expr-name expr-clause]]
+  (lib/expression query stage-number expr-name expr-clause))
+
+(defmethod before-and-after :expression [before after [_expression stage-number expr-name _expr-clause]]
+  (let [before-exprs (lib/expressions before stage-number)
+        after-exprs  (lib/expressions after stage-number)]
+    (testing "adding an expression"
+      (testing "adds it to the expressions list"
+        (is (= (inc (count before-exprs))
+               (count after-exprs))))
+      (testing "adds it to visible columns"
+        (is (=? [{:lib/type   :metadata/column
+                  :lib/source :source/expressions
+                  :name       expr-name
+                  :id         (symbol "nil #_\"key is not present.\"")}]
+                (filter #(= (:name %) expr-name) (lib/visible-columns after stage-number))))))))
```

Comment:
Any possibility for name collision with non-expression visible columns here? Like if `expr-name` happens to be `count` or something.

Reply:
As a further flourish, if you deselect `Orders.SUBTOTAL`, we don't disambiguate it anymore either - the expression is returned as `SUBTOTAL`.

There's an unsolvable point here that came up in the new refs discussion: sometimes we want to sneakily replace one column with another and have consumers not know; other times we want to strictly identify columns rather than letting them juggle labels.

First case: We may want to replace a janky database field with a more polished expression based on it, and have all the consumers of the model be none the wiser. One common example is replacing a nullable column with `coalesce([Some Field], 0)`. We want to be able to make that update to the model without updating every question that consumes it.

Second case: The disambiguation is too dynamic, and the same name (eg. `sum_2`) can be owned by different, unrelated values. Question A has three `SUM` aggregations and Question B filters on `sum_2`, the middle one. If I delete the _first_ `SUM`, there's still a `sum` and `sum_2`, but those names now refer to the columns which were previously called `sum_2` and `sum_3` respectively.

The conclusion is that the second problem is unsolvable without a stronger sense of identity for columns and references to them; this is a goal of the new refs. We can achieve the first case with a deliberate find-replace. (As a bonus, giving all fields and columns a globally unique, permanent `:ident` makes find-replace way easier.)

Reply:
This is a great comment and a perfect example of why I like to ask "stupid" questions in reviews, because it gives the reviewee a chance to talk about the broader context of the changes, which is always enlightening (especially for a noob). Super helpful & much appreciated.

I get that's it unlikely and that name collisions are valid and the qp should handle it via renaming / uniquifying, but won't this test assertion potentially fail if there is a collision, depending on the order the columns are returned by `visibile-columns`? E.g.

```clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/expression "TAX" (lib/+ 0 0))
    (lib/visible-columns)
    (->> (filter #(= (:name %) "TAX"))
         (map (juxt :name :lib/source))))

;; (["TAX" :source/table-defaults] ["TAX" :source/expressions])
```

Maybe it's unlikely enough we don't care, but I guess we could update the filter here to also filter based on `:lib/source` to ensure it's selecting the expression column (or at least *an* expression column, since two expressions could have the same name as well). Or maybe we don't care because once the refs changes land, we'll be able to uniquely identify the expression by ident and this will go away.

Reply:
How unlikely also depends on how high we plan to crank the generator and how often we run it with a non-fixed seed, but e.g.

```Clojure
(defn- gen-expression-name-until [needle maxn]
  (loop [n 1]
    (cond
      (= needle (gen-expression-name))
      [n needle]
      (> n maxn)
      [n 'limit-exceeded]
      true
      (recur (+ n 1)))))

(repeatedly 4 #(gen-expression-name-until "TAX" 1E7))
([993659 "TAX"] [2474138 "TAX"] [13399 "TAX"] [4291065 "TAX"])
```

which more-or-less matches what I'd expect (assuming I got the math right)

```clojure
(* (- 30 3) (math/pow (count expr-name-chars) 3))
1.2812904E7
```

1 in 10 million is not exactly common, but not astronomically unlikely or out of the conceivable range of how many queries / expression clauses you might want to generate. Probably not for a single run, but over time with variable seed you'd expect it to happen.